docs/javascript-unit-tests.md

```diff
@@ -0,0 +1,712 @@
+Each function in a module should have tests set up as
+a child `describe` within the module’s own `describe`.
+Then, write a number of `it` statements in prose that describe
+how the function should respond to various kinds of input.
+Inside each `it`, invoke the function
+with the input described in the `it` statement
+and use `expect` to check that you receive the desired result.
+
+Here is a simple example from our
+[array helpers module](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/unprocessed/js/modules/util/array-helpers.js)
+(`cfgov/unprocessed/js/modules/util/array-helpers.js`):
+
+```js
+function indexOfObject( array, key, val ) {
+  let match = -1;
+
+  if ( !array.length > 0 ) {
+    return match;
+  }
+
+  array.forEach( function( item, index ) {
+    if ( item[key] === val ) {
+      match = index;
+    }
+  } );
+
+  return match;
+}
+```
+
+Tests for that function, from
+[test/unit_tests/js/modules/util/array-helpers-spec.js](https://github.com/cfpb/cfgov-refresh/blob/master/test/unit_tests/js/modules/util/array-helpers-spec.js):
+
+```js
+describe( 'indexOfObject()', () => {
+  it( 'should return -1 if the array is empty', () => {
+    array = [];
+    index = arrayHelpers.indexOfObject( array, 'foo' );
+
+    expect( index ).toBe( -1 );
+  } );
+
+  it( 'should return -1 if there is no match', () => {
+    array = [
+      { value: 'bar' },
+      { value: 'baz' }
+    ];
+    index = arrayHelpers.indexOfObject( array, 'value', 'foo' );
+
+    expect( index ).toBe( -1 );
+  } );
+
+  it( 'should return the matched index', () => {
+    array = [
+      { value: 'foo' },
+      { value: 'bar' },
+      { value: 'baz' }
+    ];
+    index = arrayHelpers.indexOfObject( array, 'value', 'foo' );
+
+    expect( index ).toBe( 0 );
+  } );
+} );
+```
+
+
+### Testing DOM manipulation
+
+[Jest](https://jestjs.io/en/), the JavaScript testing framework we use,
+[includes jsdom](https://jestjs.io/docs/en/configuration#testenvironment-string),
+which simulates a DOM environment as if you were in the browser.
+This means that we can call any
+[DOM API](https://developer.mozilla.org/en-US/docs/Web/API/Document_Object_Model)
+in our test code and observe it in the same way as we do
+in the module code itself, which acts on the browser’s DOM.
+
+As an example, let’s look at our Notification component.
+The Notification component uses a common set of markup
+with different classes and SVG icon code to style it
+as a particular kind of notification (success, warning, etc.). In
+[the component JS](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/unprocessed/js/molecules/Notification.js),
+we have this function that sets the type of a notification before displaying it:
+
+```js
+function _setType( type ) {
+  // If type hasn't changed, return.
+  if ( _currentType === type ) {
+    return this;
+  }
+
+  // Remove existing type class
+  const classList = _dom.classList;
+  classList.remove( BASE_CLASS + '__' + _currentType );
```

Comment:
Should probably use a template string (can be updated in source):

```js
 `${ BASE_CLASS }__${ _currentType }`
```

Reply:
I'll update it prospectively in this excerpt, but there are lots of things in the source file that could be template strings, so that should probably be done in a separate PR.